### PR TITLE
lsp: use severity_limit in sign and virtual text

### LINKED ISSUE #13314: [LSP Diagnostics Feature Enhancement] Add severity_limit to on_publish_diagnostics (if possible)
Hello Neovim LSP folks,

The new `vim.lsp.with` + `on_publish_diagnostics` API for tweaking LSP diagnostics is very nice. Well done.

However, I still find the *TypeScript Language Server* way too chatty with regards to issues it reports.
Neovim LSP diagnostics classifies issues into four categories: Error, Warning, Information and Hint. The *TypeScript Language Server* is showing a bunch of hints which really aren't issues, nevertheless it is clutter that I want to be rid of (at a configuration level).

It would be nice to be able to tweak Neovim LSP Diagnostics to only display issues above a certain severity limit. The function `get_line_diagnostics` already is aware of the concept of `severity_limit`, hence, it would be nice to extend that (if possible) also to `vim.lsp.with`.

For example:

```viml
vim.lsp.handlers["textDocument/publishDiagnostics"] = vim.lsp.with(
  vim.lsp.diagnostic.on_publish_diagnostics, {
    signs = true,
    underline = false,
    update_in_insert = false,
    virtual_text = {
      spacing = 2,
    },
    severity_limit = 'Warning' <<< NEW OPTION, MAYBE
  }
)
```

This theoretical example states only show errors and warning, not information and hints.

Best regards.

### PR DESCRIPTION
**Pull request by glepnir**
close #13314

### CODE REVIEW

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 1]
```diff
@@ -744,12 +744,17 @@ function M.get_virtual_text_chunks_for_line(bufnr, line, line_diags, opts)
   opts = opts or {}
   local prefix = opts.prefix or "■"
   local spacing = opts.spacing or 4
+  local severity_level = opts.severity_level or nil
```
Comment (vigoux) [inline runtime/lua/vim/lsp/diagnostic.lua:747]: This or is not useful.

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 2]
```diff
@@ -744,12 +744,17 @@ function M.get_virtual_text_chunks_for_line(bufnr, line, line_diags, opts)
...
 
   -- Create a little more space between virtual text and contents
   local virt_texts = {{string.rep(" ", spacing)}}
 
   for i = 1, #line_diags - 1 do
-    table.insert(virt_texts, {prefix, virtual_text_highlight_map[line_diags[i].severity]})
+    if severity_level == nil then
```
Comment (vigoux) [inline runtime/lua/vim/lsp/diagnostic.lua:753]: This could be

```suggestion
    if not severity then
```

Comment (glepnir) [reply runtime/lua/vim/lsp/diagnostic.lua:753]: aha right.

File: runtime/lua/vim/lsp/diagnostic.lua (hunk 1 above)
Comment (glepnir) [reply runtime/lua/vim/lsp/diagnostic.lua]: done

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 3]
```diff
@@ -744,12 +744,17 @@ function M.get_virtual_text_chunks_for_line(bufnr, line, line_diags, opts)
...
 
   -- Create a little more space between virtual text and contents
   local virt_texts = {{string.rep(" ", spacing)}}
 
   for i = 1, #line_diags - 1 do
-    table.insert(virt_texts, {prefix, virtual_text_highlight_map[line_diags[i].severity]})
+    if not severity_limit then
+      table.insert(virt_texts, {prefix, virtual_text_highlight_map[line_diags[i].severity]})
+    else
+      table.insert(virt_texts, {prefix, virtual_text_highlight_map[line_diags[i][severity_level]]})
```
Comment (vigoux) [inline runtime/lua/vim/lsp/diagnostic.lua:756]: Is that a typo ?
```suggestion
      table.insert(virt_texts, {prefix, virtual_text_highlight_map[line_diags[i][severity_limit]]})
```

File: runtime/lua/vim/lsp/diagnostic.lua [hunk 4]
```diff
@@ -744,12 +744,17 @@ function M.get_virtual_text_chunks_for_line(bufnr, line, line_diags, opts)
...
   local prefix = opts.prefix or "■"
   local spacing = opts.spacing or 4
+  local severity_limit = opts.severity_limit
 
   -- Create a little more space between virtual text and contents
   local virt_texts = {{string.rep(" ", spacing)}}
 
   for i = 1, #line_diags - 1 do
-    table.insert(virt_texts, {prefix, virtual_text_highlight_map[line_diags[i].severity]})
+    if not severity_limit then
```
Comment (vigoux) [inline runtime/lua/vim/lsp/diagnostic.lua:753]: You could even invert those two branches for readability.

File: runtime/lua/vim/lsp/diagnostic.lua (hunk 3 above)
Comment (glepnir) [reply runtime/lua/vim/lsp/diagnostic.lua:756]: Sorry for my sloppyness, I submitted pr in a hurry just now, now I have time

**Review by windwp — CHANGES_REQUESTED** [review]
@glepnir It is wrong you are not testing it
You should implement like this
https://github.com/neovim/neovim/blob/e049a4ba056a778a58fddde7b9b43e9071f51f8c/runtime/lua/vim/lsp/diagnostic.lua#L378-L388

**Comment by glepnir** [comment]
if `seviery_limit` is string type this is ok. if is number this is wrong , I ignore it's number

**Pull request closed without merging**